### DISCUSSION TITLE
What if the database connection pool is not enough?

### DISCUSSION THREAD

**Discussion by elinzzyl** [fastapi/fastapi#12179] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, HTTPException
from sqlalchemy.orm import Session

from . import crud, models, schemas
from .database import SessionLocal, engine

models.Base.metadata.create_all(bind=engine)

app = FastAPI()


# Dependency
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@app.post("/users/", response_model=schemas.User)
async def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
    db_user = crud.get_user_by_email(db, email=user.email)
    if db_user:
        raise HTTPException(status_code=400, detail="Email already registered")
    return crud.create_user(db=db, user=user)

-----------------------------------------------------
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine(
    "mysql+pymysql://username:password@localhost:port/test", pool_size=20, max_overflow=0, pool_pre_ping=True
)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
```

### Description

* The above code comes from this link： https://fastapi.tiangolo.com/tutorial/sql-databases/#main-fastapi-app
* We are creating the database session before each request in the dependency with yield, and then closing it afterwards.
* If I change `create_user` to async def and initiate multiple requests at the same time, the maximum value of the connection pool will be exceeded, and an error of insufficient connection pool will be reported.
* So the question is: 
* Does writing something like db: Session = Depends(get_db) limit the number of concurrent requests?
* Is it better to post-acquire the database connection?

### Operating System

Linux

### FastAPI Version

0.87.0

### Pydantic Version

1.10.8

### Python Version

Python 3.8.10

**Comment by YuriiMotov** [comment, marked as answer]
If you are using standard pool, you will probably be able to set the `pool` limits using the parameters `pool_size` (by default is 5) and `max_overflow` (by default is 10): https://docs.sqlalchemy.org/en/20/core/engines.html#

But be careful, at some point you can reach the limit of your server.

> Does writing something like db: Session = Depends(get_db) limit the number of concurrent requests?

It's SQLAlchemy's pool that does it when you call `SessionLocal` (in this case).

> Is it better to post-acquire the database connection?

If you are talking about establishing new connection every time you need instead of using connection pool, then it's not very efficient way, because it takes time to establish new connection, and by using connection pool you can save this time.
Better to keep using connection pool, but set `max_overflow` to high enough value. In this case SQLAlchemy will keep some connections open and reuse them, but if it's not enough it will open new temporary connection
